src/tracing/esp32_diagnostic_trace/DiagnosticTracing.cpp

```diff
@@ -0,0 +1,187 @@
+
+/*
+ *
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <algorithm>
+#include <esp_err.h>
+#include <esp_log.h>
+#include <tracing/esp32_diagnostic_trace/Counter.h>
+#include <tracing/esp32_diagnostic_trace/DiagnosticTracing.h>
+
+namespace chip {
+namespace Tracing {
+namespace Diagnostics {
+
+namespace Utils {
+// Implements a murmurhash with 0 seed.
+uint32_t MurmurHash(const void * key)
+{
+    const uint32_t kMultiplier = 0x5bd1e995;
+    const uint32_t kShift      = 24;
+    const unsigned char * data = (const unsigned char *) key;
+    uint32_t hash              = 0;
+
+    while (*data)
+    {
+        uint32_t value = *data++;
+        value *= kMultiplier;
+        value ^= value >> kShift;
+        value *= kMultiplier;
+        hash *= kMultiplier;
+        hash ^= value;
+    }
+
+    hash ^= hash >> 13;
+    hash *= kMultiplier;
+    hash ^= hash >> 15;
+
+    if (hash == 0)
+    {
+        ChipLogError(DeviceLayer, "MurmurHash resulted in a hash value of 0");
+    }
+
+    return hash;
+}
+} // namespace Utils
+
+constexpr size_t kPermitListMaxSize = CONFIG_MAX_PERMIT_LIST_SIZE;
+using HashValue                     = uint32_t;
+using namespace Utils;
```

Comment:
why declare a namespace to use it right away? Anonymous namespace without using seems better.